src/components/MultiSelect/MultiSelect.tsx

```diff
@@ -1,138 +1,476 @@
-      type: "listbox";
-      onApply?: never;
-      /** The action to perform for downshift's onSelectedItemChange function. */
-      onChange: ListboxOnChange;
-      // These are props that are never allowed on the listbox type.
-      onMixedStateChange?: never;
-    }
-  | {
-      type: "dialog";
-      /** The action to perform for save/apply button of multiselect. */
-      onApply: () => void;
-      /** The action to perform on the checkbox's onChange function.  */
-      onChange: DialogOnChange;
-      /** The action to perform for a mixed state checkbox (parent checkbox). */
-      onMixedStateChange?: DialogOnChange;
-    };
-
-export type MultiSelectProps = MultiSelectCommonProps &
-  MultiSelectVariantsProps;
 
 /**
- * The `MultiSelect` component is a form input element that presents a list
- * of `Checkbox` components from which a user can make one or multiple
- * selections. Two variants of the MultiSelect component are offered, each with
- * slightly different functionality and requirements.  Because of these
- * differences, the two variants are broken out in separate stories below.
- */
+  The MultiSelect component is a customizable form input that supports multiple configurations,
+  including search functionality, checkbox options, and hierarchical structure,
+  with a parent checkbox toggling all children and dynamic styling through Chakra UI.
+*/
+
 export const MultiSelect = chakra(
   forwardRef<HTMLDivElement, React.PropsWithChildren<MultiSelectProps>>(
-    (props, ref?) => {
+    (props, _ref?) => {
       const {
         id,
         isBlockElement = false,
         isDefaultOpen = false,
+        isSearchable = false,
+        defaultItemsVisible = 5,
         items,
-        labelText,
-        onApply,
+        listOverflow = "scroll",
+        buttonText,
         onChange,
         onClear,
         onMixedStateChange,
         selectedItems,
-        type,
-        width = "default",
+        width = "full",
         ...rest
       } = props;
 
-      const commonProps = {
-        id,
+      // Create a ref to hold a reference to the accordian button, enabling us to programmatically focus it.
+      const accordianButtonRef: React.RefObject<HTMLDivElement> =
+        useRef<HTMLDivElement>();
+
+      // Control the open or closed state of the MultiSelect.
+      const DEFAULT_ITEMS_LIST_HEIGHT = "0px";
+      const MINIMUM_ITEMS_LIST_HEIGHT = "215px";
+      const MAXIMUM_ITEMS_LIST_HEIGHT = "270px";
+      const [itemsList, setItemsList] = useState(items);
+      const [viewAllLabel, setViewAllLabel] = useState("View all");
+      const [listHeight, setListHeight] = useState(MINIMUM_ITEMS_LIST_HEIGHT);
+      const [listItemsCount, setListItemsCount] = useState(defaultItemsVisible);
+
+      // Separate effect for handling listOverflow "scroll"
+      React.useEffect(() => {
+        if (listOverflow === "scroll") {
+          setListHeight(MINIMUM_ITEMS_LIST_HEIGHT);
+          if (isSearchable) {
```

Comment:
To fix a bug with storybook where the `itemsList` are not being updated correctly when the listOverflow type is changes in the controls I added this useEffect. 
```
    // Only for storybook
      React.useEffect(() => {
        setItemsList(defaultItemsList);
      }, [listOverflow]);
```
This is only an issue in storybook as in a actual use case the MultiSelect would not just randomly change the way it is displaying the options (I don't think?). This is only a temporary fix.

Reply:
No, this prop would not change randomly. If this is a limitation/bug from Storybook, then mention it in the docs and add two different static examples. I don't like the idea of having and shipping code to fix Storybook in the actual component.